### PR TITLE
fix(curriculam): Learn Functional Programming by Building a Spreadsheet - has2 not recognized as spreadsheet function

### LINKED ISSUE #53641: Learn Functional Programming by Building a Spreadsheet - `has2` not recognized as spreadsheet function
### Describe the Issue

The spreadsheet function "has2" (first defined in step 97 as of Feb 10) does not work in the spreadsheet page. I think this is because the regex in the function applyFunction doesn't recognize it as a function.
The regex:
const functionCall = /([a-z]*)\(([0-9., ]*)\)(?!.*\()/i;

I think the regex needs to change to: /([a-z2]*)\(([0-9., ]*)\)(?!.*\()/i
or: /([a-z0-9]*)\(([0-9., ]*)\)(?!.*\()/i

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-functional-programming-by-building-a-spreadsheet/step-97

### Your code

```
const functionCall = /([a-z]*)\(([0-9., ]*)\)(?!.*\()/i;


```

### Expected behavior

When a formula in the spreadsheet is:
=has2(1,2,3)
The cell updates to "true"
When formula is
=has2(4,5,6)
The cell updates to "false"

### Screenshots

_No response_

### System

- Device: [ Laptop]
- OS: [MacOS 14.3]
- Browser: [Chrome]
- Version: [120]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Nitz2611**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #53641

### CODE REVIEW

**Comment by gikf** [comment]
This will require changing steps from the point when `functionCall` is introduced (step 82) forward. However I'd suggest waiting with changing other steps until new regex is reviewed.

**Comment by Nitz2611** [comment]
@gikf Ok. Should i close this pr and open after the new regex is reviewed.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d49bfff9079d4b38df115.md [hunk 1]
```diff
@@ -135,7 +135,7 @@ const applyFunction = str => {
   const noHigh = highPrecedence(str);
   const infix = /([\d.]+)([+-])([\d.]+)/;
   const str2 = infixEval(noHigh, infix);
-  const functionCall = /([a-z]*)\(([0-9., ]*)\)(?!.*\()/i;
```
Comment (DBailey0629) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d49bfff9079d4b38df115.md:138]: @Nitz2611 I think you'll want to leave the regex unchanged except for the addition of the number set in the first capture group.

Comment (Nitz2611) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d49bfff9079d4b38df115.md:138]: done

**Comment by Nitz2611** [comment]
@gikf i have to change all the steps from step 81, where `functionCall` was introduced, right?

**Comment by jdwilkin4** [comment]
@Nitz2611 

It looks like some of automated tests are failing.
You will need to run the following command locally to see the errors and fix them. Once everything passes locally, then you can push up you changes and ready for another review

```
FCC_BLOCK='Learn Functional Programming by Building a Spreadsheet' pnpm run test:curriculum

```

**Comment by Nitz2611** [comment]
@jdwilkin4 This error is occuring while running the command you provided. How do i resolve it?
error

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d448479c8fdc8dcec868c.md
Comment (gikf) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d448479c8fdc8dcec868c.md:1]: Third test needs updating to the new pattern in this file. GitHub UI doesn't allow right now to make change suggestion at that place.

Comment (Nitz2611) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d448479c8fdc8dcec868c.md:1]: Could you please provide more details on what exactly needs to be updated in the third test? I'm not fully clear on the new pattern or structure you're referring to, and I want to make sure I understand the changes required.

Comment (gikf) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d448479c8fdc8dcec868c.md:1]: Thing is, all tests from this step should pass when checked against seed code from next step. 

Third test:
```javascript
assert.match(code, /const\s+applyFunction\s*=\s*\(?\s*str\s*\)?\s*=>\s*\{\s*const\s+noHigh\s*=\s*highPrecedence\(\s*str\s*\);?\s*const\s+infix\s*=\s*\/\(\[(?:\\d\.|\.\\d)\]\+\)\(\[(?:\+-|-\+)\]\)\(\[(?:\\d\.|\.\\d)\]\+\)\/;?\s*const\s+str2\s*=\s*infixEval\(\s*noHigh\s*\,\s*infix\s*\);?\s*const\s+functionCall\s*=\s*\/\(\[a-z\]\*\)\\\(\(\[0-9\., \]\*\)\\\)\(\?!\.\*\\\(\)\/i/);
```
Changed seed code from next step (only relevant part):
```javascript
const applyFunction = str => {
  const noHigh = highPrecedence(str);
  const infix = /([\d.]+)([+-])([\d.]+)/;
  const str2 = infixEval(noHigh, infix);
  const functionCall = /([a-z0-9]*)\(([0-9., ]*)\)(?!.*\()/i;

}
```

Right now that's failing due to `0-9` added in seed code, therefore test regex needs to be updated for it:
```diff
-assert.match(code, /const\s+applyFunction\s*=\s*\(?\s*str\s*\)?\s*=>\s*\{\s*const\s+noHigh\s*=\s*highPrecedence\(\s*str\s*\);?\s*const\s+infix\s*=\s*\/\(\[(?:\\d\.|\.\\d)\]\+\)\(\[(?:\+-|-\+)\]\)\(\[(?:\\d\.|\.\\d)\]\+\)\/;?\s*const\s+str2\s*=\s*infixEval\(\s*noHigh\s*\,\s*infix\s*\);?\s*const\s+functionCall\s*=\s*\/\(\[a-z\]\*\)\\\(\(\[0-9\., \]\*\)\\\)\(\?!\.\*\\\(\)\/i/);
+assert.match(code, /const\s+applyFunction\s*=\s*\(?\s*str\s*\)?\s*=>\s*\{\s*const\s+noHigh\s*=\s*highPrecedence\(\s*str\s*\);?\s*const\s+infix\s*=\s*\/\(\[(?:\\d\.|\.\\d)\]\+\)\(\[(?:\+-|-\+)\]\)\(\[(?:\\d\.|\.\\d)\]\+\)\/;?\s*const\s+str2\s*=\s*infixEval\(\s*noHigh\s*\,\s*infix\s*\);?\s*const\s+functionCall\s*=\s*\/\(\[a-z0-9\]\*\)\\\(\(\[0-9\., \]\*\)\\\)\(\?!\.\*\\\(\)\/i/);
```

**Comment by gikf** [comment]
Hey @Nitz2611,

sorry, I forgot there are more steps that are using that part of regex. Looks like updating tests up to step 88 will be needed.

**Comment by Nitz2611** [comment]
@gikf still some tests are failing, why?

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d467c6994f4ce0dc416a4.md [hunk 2]
```diff
@@ -14,7 +14,7 @@ Use the ternary operator to turn your `.hasOwnProperty()` call into the conditio
...
+assert.match(code, /const\s+applyFunction\s*=\s*\(?\s*str\s*\)?\s*=>\s*\{\s*const\s+noHigh\s*=\s*highPrecedence\(\s*str\s*\);?\s*const\s+infix\s*=\s*\/\(\[(?:\\d\.|\.\\d)\]\+\)\(\[(?:\+-|-\+)\]\)\(\[(?:\\d\.|\.\\d)\]\+\)\/;?\s*const\s+str2\s*=\s*infixEval\(\s*noHigh\s*,\s*infix\s*\);?\s*const\s+functionCall\s*=\s*\/\(\[a-z0-9\]\*\)\\\(\(\[0-9\., \]\*\)\\\)\(\?!\.\*\\\(\)\/i;?\s*const\s+toNumberList\s*=\s*\(?\s*args\s*\)?\s*=>\s*args\.split\(\s*('|"|`),\1\s*\)\.map\(\s*parseFloat\s*\);?\s*const\s+apply\s*=\s*\(\s*fn\s*,\s*args\s*\)\s*=>\s*spreadsheetFunctions\s*\[\s*fn\.toLowerCase\(\s*\)\s*\]\s*\(\s*toNumberList\(\s*args\s*\)\s*\);?\s*return\s+str2\.replace\(\s*functionCall\s*,\s*\(\s*match\s*,\s*fn\s*,\s*args\s*\)\s*=>\s*spreadsheetFunctions\.hasOwnProperty\(\s*fn\.toLowerCase\(\s*\)\s*\)\s*\?/);
 ```
 
 If the ternary condition is true, your callback function should return the result of calling `apply()`.
```
Comment (gikf) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/646d467c6994f4ce0dc416a4.md:20]: One more regex to change for this hint. (GitHub UI doesn't let to make suggestion at the line with actual regex)

**Comment by Nitz2611** [comment]
hey @gikf I think now it's passing all the tests. Thankyou for your assistance, and I truly appreciate all the time and effort you invested in helping me through the process.

**Review by gikf — APPROVED** [review]

**Review by Sembauke — APPROVED** [review]

**Pull request merged**